Add argocd appsets based application samples

abhijeet219 opened the pull request:
The PR adds argocd applicationsets based application samples.

**Testing Summary:**
Verified ApplicationSet functionality across all workload types:

Deployment workload: 
`kubectl apply -k applicationset/deployment-k8s-regional-rbd/`

KubeVirt VM workloads:
```
kubectl apply -k applicationset/kubevirt/vm-dv-k8s-regional/
kubectl apply -k applicationset/kubevirt/vm-dv-odr-regional/
kubectl apply -k applicationset/kubevirt/vm-dvt-k8s-regional/
kubectl apply -k applicationset/kubevirt/vm-dvt-odr-regional/
kubectl apply -k applicationset/kubevirt/vm-pvc-k8s-regional/
kubectl apply -k applicationset/kubevirt/vm-pvc-odr-regional/
```

Results:
- All ApplicationSet resources created successfully in argocd namespace
- ArgoCD Applications generated correctly for each workload type
- Placement successfully selected dr1 cluster using predicate-based targeting
- No deprecation warnings after updating from commonLabels to labels

**Steps followed for testing:**
1. Initial Setup
Ensured ArgoCD RBAC permissions for OCM integration by applying the required ClusterRole and ClusterRoleBinding.
Verified OCM hub cluster context was active
2. Applied the ApplicationSet Configuration
`kubectl apply -k applicationset/deployment-k8s-regional-rbd/`
3. Verified ApplicationSet Creation
`kubectl get applicationsets -n argocd`
Expected: deployment-rbd-appset should be listed
4. Checked Generated Applications
`kubectl get applications -n argocd`
Expected: deployment-rbd-appset-dr1 should be created and eventually show Synced status
5. Confirmed Workload Deployment on Target Cluster
`kubectl get pods -n deployment-rbd --context=dr1`
Expected: busybox pod should be running on the dr1 cluster

abhijeet219 committed:
Add argocd applicationset samples

Replacing the subscriptions based application samples with argocd
applicationsets based application samples.

Signed-off-by: Abhijeet Shakya <abhijeetshakya21@gmail.com>

Add predicate to placement, modify it using kustomize

Signed-off-by: Abhijeet Shakya <abhijeetshakya21@gmail.com>

rakeshgm commented:
@abhijeet219 Is this effecting in the way we deploy workloads from in `drenv`. do we need any changes there ?

ShyamsundarR commented on applicationset/base/applicationset.yaml:
```diff
@@ -0,0 +1,30 @@
+      repoURL: https://github.com/RamenDR/ocm-ramen-samples.git
+      revision: main
+      directories:
+        - path: workloads/*/*
```
Should this also be PATH instead of `workloads/*/*`?

abhijeet219 replied:
Addressed.

ShyamsundarR commented on applicationset/deployment-k8s-regional-rbd/namespace.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
This will be created on the hub, do we need the app namespace on the hub cluster?

abhijeet219 replied:
Agree, removed the namespace.

ShyamsundarR commented on applicationset/base/applicationset.yaml:
```diff
@@ -0,0 +1,30 @@
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: ramen-appset-sample
```
As all AppSets would get deployed to the same namespace `arocd` we need to kustomize the name for the AppSet as well in each overlay directory?

abhijeet219 replied:
Fixed.

ShyamsundarR reviewed (commented):
We still need a Placement and ManagedClusterSetBinding resource for the AppSet as well, afaict.

ShyamsundarR commented:
In the Ramen repo, the following are affected:
- test/configs/deployment-k8s-regional-rbd.yaml (and other configs)
- test/basic-test/config.yaml

The way to fix it would be to split this into 2 PRs, first one adding the AppSet, and once that is merged amending Ramen code as above to use AppSet, and then the next PR in samples to remove the Subscriptions.

abhijeet219 committed:
Replace deprecated commonLabels with labels

abhijeet219 committed:
Update README with applicationset sample commands

Since we are removing the subscription base application deployment.
And replacing that with applicationset based application samples,
updating README to accomodate the changes.

Signed-off-by: Abhijeet Shakya <abhijeetshakya21@gmail.com>

nirs reviewed (commented):
Please share output of all commands comparing what we had for subscription and applicationset. We want to make sure the names of the resources make sense and are like the previous names.

abhijeet219 commented on applicationset/base/applicationset.yaml:
```diff
@@ -0,0 +1,36 @@
+          - path: WORKLOAD_PATH
+      - list:
+          elements:
+          - name: PLACEMENT_SELECTED_CLUSTER
```
I tried to update the configMap manually with the cluster details from PlacementDecision, but it became very complicated and doesn't seem to be the right way to do it.

We have 2 approaches now:
1. Using Not-in-cluster generator: which deploys to all the clusters apart from hub.
2. Directly specify the cluster we want to deploy to, in the overlay.

Suggestions?
